crates/wasi-http/src/body.rs

```diff
@@ -0,0 +1,513 @@
+
+        TrailersReady(self)
+    }
+}
+
+pub type HyperBody = BoxBody<Bytes, Infallible>;
+
+pub struct HostOutgoingBody {
+    pub body_output_stream: Option<Box<dyn HostOutputStream>>,
+    pub trailers_sender: Option<tokio::sync::oneshot::Sender<hyper::HeaderMap>>,
+}
+
+impl HostOutgoingBody {
+    pub fn new() -> (Self, HyperBody) {
+        use http_body_util::BodyExt;
+        use hyper::{
+            body::{Body, Frame},
+            HeaderMap,
+        };
+        use std::future::Future;
+        use std::task::{Context, Poll};
+        use tokio::sync::oneshot::error::RecvError;
+        struct BodyImpl {
+            body_receiver: mpsc::Receiver<Bytes>,
+            trailers_receiver: Option<oneshot::Receiver<HeaderMap>>,
+        }
+        impl Body for BodyImpl {
+            type Data = Bytes;
+            type Error = Infallible;
+            fn poll_frame(
+                mut self: Pin<&mut Self>,
+                cx: &mut Context<'_>,
+            ) -> Poll<Option<Result<Frame<Self::Data>, Self::Error>>> {
+                match self.as_mut().body_receiver.poll_recv(cx) {
+                    Poll::Pending => Poll::Pending,
+                    Poll::Ready(Some(frame)) => Poll::Ready(Some(Ok(Frame::data(frame)))),
+
+                    // This means that the `body_sender` end of the channel has been dropped.
+                    Poll::Ready(None) => {
+                        if let Some(mut trailers_receiver) = self.as_mut().trailers_receiver.take()
+                        {
+                            match Pin::new(&mut trailers_receiver).poll(cx) {
+                                Poll::Pending => {
+                                    self.as_mut().trailers_receiver = Some(trailers_receiver);
+                                    Poll::Pending
+                                }
+                                Poll::Ready(Ok(trailers)) => {
+                                    Poll::Ready(Some(Ok(Frame::trailers(trailers))))
+                                }
+                                Poll::Ready(Err(RecvError { .. })) => Poll::Ready(None),
+                            }
+                        } else {
+                            Poll::Ready(None)
+                        }
+                    }
+                }
+            }
+        }
+
+        let (body_sender, body_receiver) = mpsc::channel(1);
+        let (trailers_sender, trailers_receiver) = oneshot::channel();
+        let body_impl = BodyImpl {
+            body_receiver,
+            trailers_receiver: Some(trailers_receiver),
+        }
+        .boxed();
+        (
+            Self {
+                // TODO: this capacity constant is arbitrary, and should be configurable
+                body_output_stream: Some(Box::new(BodyWriteStream::new(1024 * 1024, body_sender))),
+                trailers_sender: Some(trailers_sender),
+            },
+            body_impl,
+        )
+    }
+}
+
+// copied in from preview2::write_stream
```

Comment:
This is the part of the PR that gives me the most pause. Not necessarily the duplication really but moreso the layers of abstraction. This feels like an extra and not necesarily required layer of buffering that otherwise may not be needed. 

So question for you. Reading over [some docs](https://docs.rs/hyper/1.0.0-rc.4/hyper/body/trait.Body.html) I see:

```rust
impl<S, D, E> hyper::Body for http_body_util::StreamBody<S>
where
    S: futures_util::Stream<Item = Result<hyper::Frame<D>, E>>,
    D: Buf,
```

The writer below is more-or-less a simplistic `mpsc::Sender<Bytes>` where the paried end of `mpsc::Receiver<Bytes>` seems like it would fit the above signature perfectly. In that sense it seems like `impl Body for BodyImpl` above this and most of the below could could be deleted in favor of that. One weird part is flushes though but the `Body` trait doesn't seem to have a `flush`-method equivalent so there's not much I think that can be done about that.

I'm curious though if y'all dug into this and selected the current strategy due to various characteristics or something like that?

Reply:
Yes, theres probably a better implementation possible here if we simplify the write budget down to a single frame with some bound. We planned to go back and revisit that at some later point, this was the hack of some code we understood that seemed to work good enough for the moment.